### PR TITLE
Protect pulse generation timing by disabling interrupts for longer

### LINKED ISSUE #27115: [BUG] `SysTick_Callback` interferes with timing critical code - causes speed jitter - LPC176x
### Did you test the latest `bugfix-2.1.x` code?

Yes, and the problem still exists.

### Bug Description

`SysTick_Callback()` should not be used as doing so causes timing jitter for *all interrupts*. `SysTick_Callback()` is an api from the `pio-framework-arduino-lpc176x`.

Here's the relevant section from [`pio-framework-arduino-lpc176x/cores/arduino/main.cpp`](https://github.com/p3p/pio-framework-arduino-lpc176x/blob/master/cores/arduino/main.cpp):

```C++
 __attribute__ ((weak)) void SysTick_Callback() {}
// [..]
volatile uint64_t _millis;
// [..]
    NVIC_SetPriority(SysTick_IRQn, NVIC_EncodePriority(0, 0, 0));
// [..]
  void SysTick_Handler(void) {
    ++_millis;
    SysTick_Callback();
  }
```

`0` is the highest priority level which means `SysTick_Handler` is going to run regardless of anyone. This would be absolutely fine if it was just incrementing `_millis`. But it's not.

`SysTick_Callback()` running as part of SysTick_IRQn and at the highest possible priority puts it ahead of every other interrupt: stepper isr, uart, serial, pwm, watchdog, you name it. And it runs every millisecond.

The main problematic `SysTick_Callback()` is from `Marlin/src/HAL/LPC1768/HAL.cpp`:
``` C++
void SysTick_Callback() { disk_timerproc(); }
```

Where `disk_timerproc()` is actually from `framework-arduino-lpc176x/system/CMSIS/lib/chanfs/mmc_ssp.c`.

## Issue
A logic analyzer capture of step/dir pins best shows the issue:
jitter
(Note in this capture the stepper isr was at the priority `0`)

You can see speed jitter equivalent to 6000mm/s^2 or more throughout, even at a somewhat leisurely 100mm/s movement speed. In my testing, I've disabled all features I could think of incl. lcd and watchdog and increased the stepper isr priority to highest (`0`), but still jitter.

## Root cause
To confirm that it's `SysTick_Callback()` I've commented it out completely, including definition and use. This is the result:
jitter_no
(Note in this capture the stepper isr was at the priority `0`)

The slightly strange part is that without the `disk_timerproc();` call, reading from/writing to SD card still seems to work. So It's definitely not crucial. Is it even needed? Every ms? / When is it needed?

## One more

`Marlin/src/lcd/extui/mks_ui/tft_lvgl_configuration.cpp` (unused in my build) also defines a `SysTick_Callback()`.

## pio-framework-arduino-lpc176x

I've also opened [pio-framework-arduino-lpc176x/issues/49](https://github.com/p3p/pio-framework-arduino-lpc176x/issues/49).

## Impact

It makes ADAPTIVE_STEP_SMOOTHING seem lees effective. ADAPTIVE_STEP_SMOOTHING reduces a larger source of speed jitter, namely bresenham aliasing for the minor axes.
It may interfere with communication like serial and spi, depending on timing sensitivity or implementation.

### Bug Timeline

Old issue, maybe 2019. Git blame points to `SysTick_Callback` added in 60cb36bef3644640f2eb1c9d2b30189e41e81ef2 "(grafted)"

### Expected behavior

No significant isr jitter.

### Actual behavior

Interrupt jitter.

### Steps to Reproduce

Run this simple/X-axis only gcode and look at the step/dir pins with a logic analyzer (at >=12Mhz sample rate).

```gcode
M92 X80 Y80 ; 80 steps per mm
G92 X0 Y0
G4 P550 ; configurable delay
M205 X10 Y10
M204 S5000
G0 F7500 X60
G0 F5400 X120
G0 F5700 X180
G0 F6000 X240
G0 F6000 X0
```

### Version of Marlin Firmware

Last sync: 06762db050399137796bbd192b94ed9da4a77e11 [cron] Bump distribution date (2024-05-18)

### Printer model

Customized Ender-3

### Electronics

SKR v1.4 Turbo but all LPC176x are affected

### LCD/Controller

_No response_

### Other add-ons

_No response_

### Bed Leveling

None

### Your Slicer

None

### Host Software

SD Card (headless)

### Don't forget to include

- [ ] A ZIP file containing your `Configuration.h` and `Configuration_adv.h`.

### Additional information & file uploads

(In lieu of configuration zip file I've root caused to specific lines of code.)

### PR DESCRIPTION
**Pull request by mh-dm**
### Description

Protect pulse generation timing by disabling interrupts within `Stepper::isr()` for longer.

This is a mitigation for the stepper isr jitter seen in #27115 and other sources of stepper isr jitter on other platforms. On microcontrollers with ISR prioritization, note that only higher priority ISRs can ever interrupt the stepper ISR.

### Benefits

More consistent pulse timings. I've tested on LPC176x and here's some before/after logic analyzer captures:

Before:
jitter-interrupt-before

After (edit, recaptured due to forgotten local changes):
jitter-interrupt-after

Basically no improvements in actual printing quality/noise unless `ADAPTIVE_STEP_SMOOTHING` is enabled since `ADAPTIVE_STEP_SMOOTHING` reduces a larger source of speed jitter, namely bresenham aliasing for the minor axes.

### Related Issues

Mostly mitigates #27115

### CODE REVIEW

**Comment by rondlh** [comment]
Would this affect "INTERRUPT_FEATURE", causing inaccurate probing?

I just did a quick repeatability test, 6x M48 for each case, and found no negative impact on probing accuracy (STM32F407 motherboard).

**Comment by thinkyhead** [comment]
I believe I resolved the merge conflict correctly. Anything that improves on timing is welcome. This should not negatively impact any of the main loop and UI processes. And it seems to cope with the possibility of affecting UART interrupts. Does it also fully resolved the linked bug report?

**Comment by mh-dm** [comment]
> Would this affect "INTERRUPT_FEATURE", causing inaccurate probing?

It won't affect it negatively. Pulse generation is really really fast and the code change is for the most time critical section at the start of the stepper isr. If the endstop/probe interrupt triggers while the stepper isr is just starting it will get delayed by at most 1 pulse/microstep (2.5 microns at the common 400steps/mm in Z). Worrying about that is moot since if micron accuracy matters then probing is done slowly - couple mm/s or ~1 microstep per millisecond whereas pulse generation is orders of magnitude faster. The probing is done slowly not due to processing latency but due to mechanical "settling" time - in particular vibrations that can affect probing by much more than 2.5 microns.

Actually, better pulse timing means slightly smoother motion thus less vibration so in theory we could probe faster at same accuracy? :)

> I just did a quick repeatability test, 6x M48 for each case, and found no negative impact on probing accuracy (STM32F407 motherboard).

Thank you for testing!

> I believe I resolved the merge conflict correctly

Looks good to me.

> Does it also fully resolved the linked bug report?

Not fully but it does mitigate it. I think we can call it good enough for now.
